### PR TITLE
Introduce limit on gasLimit

### PR DESCRIPTION
**Pull request by JustynaBroniszewska**
This PR contains #1146

### CODE REVIEW

File: universal-login-contracts/contracts/wallet/Executor.sol [hunk 1]
```diff
@@ -26,6 +26,10 @@ contract Executor {
         return 21000;
     }
 
+    function gasLimitMargin() public pure returns(uint) {
```
Comment (marekkirejczyk) [inline universal-login-contracts/contracts/wallet/Executor.sol]: `gasLimitBase`

File: universal-login-contracts/contracts/wallet/Executor.sol [hunk 2]
```diff
@@ -26,6 +26,10 @@ contract Executor {
...
+        return 40000;
```
Comment (marekkirejczyk) [inline universal-login-contracts/contracts/wallet/Executor.sol]: gasLimitMargin = 40000 - 21000 = 19000

File: universal-login-contracts/lib/estimateGas.ts [hunk 3]
```diff
@@ -25,3 +25,11 @@ export const estimateGasDataFromSignedMessage = (signedMessage: SignedMessage) =
   const txdata = encodeDataForExecuteSigned(copySignedMessage);
   return computeGasData(txdata);
 };
+
+export const estimateGasLimit = (gasLimitExecution: string | number | utils.BigNumberish, gasData: string | number | utils.BigNumberish) =>
+  utils.bigNumberify(gasLimitExecution).add(gasData).add('29000');
```
Comment (marekkirejczyk) [inline universal-login-contracts/lib/estimateGas.ts]: Did you mean 30k? 21000 + 9000?

File: universal-login-contracts/lib/estimateGas.ts [hunk 4]
```diff
@@ -25,3 +25,11 @@ export const estimateGasDataFromSignedMessage = (signedMessage: SignedMessage) =
   const txdata = encodeDataForExecuteSigned(copySignedMessage);
   return computeGasData(txdata);
 };
+
+export const estimateGasLimit = (gasLimitExecution: string | number | utils.BigNumberish, gasData: string | number | utils.BigNumberish) =>
```
Comment (marekkirejczyk) [inline universal-login-contracts/lib/estimateGas.ts]: rename `calculateFinalGasLimit`?

Comment (marekkirejczyk) [reply universal-login-contracts/lib/estimateGas.ts]: From ethers:
```ts
export type BigNumberish = BigNumber | string | number | Arrayish;
```

Unnecessary complex types

File: universal-login-contracts/test/contracts/Executor/Executor-void.ts [hunk 5]
```diff
@@ -23,7 +23,8 @@ describe('Void Executor', () => {
   it('execute signed fails', async () => {
     signature = [];
     message = {...transferMessage, from: walletContractWithZeroKey.address};
-    await expect(walletContractWithZeroKey.executeSigned(...getExecutionArgs(message), signature, TEST_PAYMENT_OPTIONS_NO_GAS_TOKEN))
+    const gasLimit = estimateGasLimit(message.gasLimitExecution, estimateGasDataForNoSignature(message));
```
Comment (marekkirejczyk) [inline universal-login-contracts/test/contracts/Executor/Executor-void.ts]: Shouldn't we always assume at least on signature?

File: universal-login-contracts/lib/estimateGas.ts [hunk 6]
```diff
@@ -25,3 +25,11 @@ export const estimateGasDataFromSignedMessage = (signedMessage: SignedMessage) =
...
+  utils.bigNumberify(gasLimitExecution).add(gasData).add('29000');
+
+export const calculatePaymentOptions = (msg: SignedMessagePaymentOptions) =>
+  ({gasLimit: estimateGasLimit(msg.gasLimitExecution, msg.gasData)});
+
+export const estimateGasDataForNoSignature = (unsignedMessage: UnsignedMessage) => estimateGasDataFromSignedMessage({...unsignedMessage, signature: '0x0'});
```
Comment (marekkirejczyk) [inline universal-login-contracts/lib/estimateGas.ts]: move to test helper

File: universal-login-contracts/contracts/wallet/Executor.sol [hunk 7]
```diff
@@ -49,6 +53,7 @@ contract Executor {
     {
         uint256 startingGas = gasleft();
         require(gasLimitExecution <= startingGas, "Gas limit too low");
```
Comment (marekkirejczyk) [inline universal-login-contracts/contracts/wallet/Executor.sol]: Relayer set gas limit to low

File: universal-login-contracts/contracts/wallet/Executor.sol [hunk 8]
```diff
@@ -49,6 +53,7 @@ contract Executor {
...
+        require(startingGas <= gasLimitExecution.add(gasLimitMargin()), "Gas limit too high");
         require(signatures.length != 0, "Invalid signatures");
         require(signatures.length >= requiredSignatures * 65, "Not enough signatures");
         bytes32 messageHash = calculateMessageHash(address(this), to, value, data, lastNonce, gasPrice, gasToken, gasLimitExecution, gasData);
```
Comment (marekkirejczyk) [inline universal-login-contracts/contracts/wallet/Executor.sol]: Relayer set gas limit to high

File: universal-login-contracts/test/helpers/argumentsEncoding.ts [hunk 9]
```diff
@@ -42,8 +43,10 @@ export const setupUpdateMessage = async (proxyAsWalletContract: Contract, newWal
     data: updateData,
     value: 0,
     nonce: await proxyAsWalletContract.lastNonce(),
-    gasLimit: DEFAULT_GAS_LIMIT.toString(),
+    gasLimit: '200000',
```
Comment (marekkirejczyk) [inline universal-login-contracts/test/helpers/argumentsEncoding.ts]: Shouldn't be calculated?

File: universal-login-contracts/contracts/wallet/Executor.sol (hunk 7 above)
Comment (JustynaBroniszewska) [reply universal-login-contracts/contracts/wallet/Executor.sol]: ✅

File: universal-login-contracts/contracts/wallet/Executor.sol (hunk 8 above)
Comment (JustynaBroniszewska) [reply universal-login-contracts/contracts/wallet/Executor.sol]: ✅

**Pull request merged**